packages/tldraw/src/lib/shapes/shared/PathBuilder.tsx

```diff
@@ -0,0 +1,674 @@
+import {
+	assert,
+	clamp,
+	exhaustiveSwitchError,
+	Geometry2dOptions,
+	getPerfectDashProps,
+	Group2d,
+	invLerp,
+	lerp,
+	Polygon2d,
+	Polyline2d,
+	rng,
+	toDomPrecision,
+	Vec,
+	VecLike,
+	VecModel,
+} from '@tldraw/editor'
+import { SVGProps } from 'react'
+
+function getVerticesCountForLength(length: number, spacing = 20) {
+	return Math.max(8, Math.ceil(length / spacing))
+}
+
+export interface BasePathBuilderOpts {
+	strokeWidth: number
+	forceSolid?: boolean
+	props?: SVGProps<SVGPathElement & SVGGElement>
+}
+
+export interface SolidPathBuilderOpts extends BasePathBuilderOpts {
+	style: 'solid'
+}
+
+export interface DashedPathBuilderOpts extends BasePathBuilderOpts {
+	style: 'dashed' | 'dotted'
+	snap?: number
+	end?: 'skip' | 'outset' | 'none'
+	start?: 'skip' | 'outset' | 'none'
+	lengthRatio?: number
+}
+
+export interface DrawPathBuilderOpts extends BasePathBuilderOpts {
+	style: 'draw'
+	randomSeed: string
+	offset?: number
+	roundness?: number
+	passes?: number
+}
+
+export type PathBuilderOpts = SolidPathBuilderOpts | DashedPathBuilderOpts | DrawPathBuilderOpts
+
+interface SegmentOpts {
+	offset?: number
+	roundness?: number
+}
+
+interface LineOpts extends SegmentOpts {
+	geometry?: Omit<Geometry2dOptions, 'isClosed'> | false
+}
+
+interface MoveToPathBuilderSegment {
+	type: 'moveTo'
+	x: number
+	y: number
+	opts?: LineOpts
+}
+
+interface LineToPathBuilderSegment {
+	type: 'lineTo'
+	x: number
+	y: number
+	opts?: SegmentOpts
+}
+
+interface ArcToPathBuilderSegment {
+	type: 'arcTo'
+	radius: number
+	largeArcFlag: boolean
+	sweepFlag: boolean
+	x: number
+	y: number
+	opts?: SegmentOpts
+}
+
+type PathBuilderSegment = LineToPathBuilderSegment | ArcToPathBuilderSegment
+
+interface PathBuilderLine {
+	initial: MoveToPathBuilderSegment
+	segments: PathBuilderSegment[]
+	closed: boolean
+}
+
+export class PathBuilder {
```

Comment:
makes me think we should call this Turtle (with apologies to Logo) :P

Reply:
heh, could do! that `Path` part at least comes somewhat from the webs [`Path2D`](https://developer.mozilla.org/en-US/docs/Web/API/Path2D) which has some similarities.